### ISSUE TITLE
needless_borrow should trigger when taking a reference to Option<&_>

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#11326]
### Lint Name

needless_borrow

### Reproducer

I tried this code:

```rust
struct Object;

impl Object {
    fn str_getter(&self) -> Option<&str> {
        None
    }
}

fn main() {
    let o = Object;
    if let Some(s) = &o.str_getter() {
        println!("{s}");
    }
}
```

I expected to see this happen: `needless_borrow` triggering for `&o.str_getter()`

Instead, this happened: nothing

### Version

Tested on Rust playground, hence less info. Let me know if you need me to reproduce locally.

```text
Nightly version: 1.73.0-nightly
(2023-08-10 439d066bcf9496b1b8c8)
```

**Comment by samueltardieu** [comment]
Just to make sure I understand the issue at hand, on which part of code would the `needless_borrow` lint apply?

If this is about the `&` in `&o.str_getter()`, removing it would change the type of `s` from `&&str` to `&str` due to match ergonomics. Even if it would work here as `Display` is implemented the same way on `&&str` and `&str`, it is not equivalent if `s` is later used in a larger expression.

Of course, in `(&o).str_getter()`, the lint would trigger, as this borrow would indeed be needless.

**Comment by jplatte** [comment]
I didn't think of that. I guess involving extra analysis on whether the bound variable from the `if let` needs to be a double ref would be become pretty complex pretty fast. Feel free to close this.
